File: bench/loadtest.mjs
// Low-overhead open-loop HTTP generator. Uses raw JSONL lines for cache-hit requests.
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { performance } from "node:perf_hooks";
import { createHash, randomUUID } from "node:crypto";

const argv = Object.fromEntries(process.argv.slice(2).reduce((pairs, value, index, all) => {
  if (value.startsWith("--")) pairs.push([value.slice(2), all[index + 1]]);
  return pairs;
}, []));
const url = argv.url ?? "http://127.0.0.1:8000/audit";
const payloadPath = argv.payloads ?? "data/corpus/ctg-randomized-2000.jsonl";
const scenario = argv.scenario;
if (!["cold", "mixed", "warm"].includes(scenario)) throw new Error("--scenario cold|mixed|warm is required");
const requests = Number(argv.requests ?? 20000);
const rate = Number(argv.rate ?? 200);
const concurrency = Number(argv.connections ?? 128);
const hitRatio = Number(argv["hit-ratio"] ?? 0.9);
const output = argv.output;
if (!output) throw new Error("--output is required");
const runId = argv["run-id"] ?? randomUUID();
const raw = fs.readFileSync(payloadPath, "utf8").trim();
const payloadSha256 = createHash("sha256").update(fs.readFileSync(payloadPath)).digest("hex");
const payloads = payloadPath.endsWith(".jsonl")
  ? raw.split(/\r?\n/).filter(Boolean)
  : (Array.isArray(JSON.parse(raw)) ? JSON.parse(raw) : [JSON.parse(raw)]).map(JSON.stringify);
const latencies = [];
const serverLatencies = [];
const errors = [];
const cacheHeaders = {};

function withNonce(payload, nonce) {
  const value = JSON.parse(payload);
  value.metadata = { ...(value.metadata ?? {}), benchmark_nonce: nonce };
  return JSON.stringify(value);
}

function bodyFor(index) {
  const base = payloads[index % payloads.length];
  if (scenario === "cold") return withNonce(base, `${runId}:cold:${index}`);
  if (scenario === "mixed") {
    const targetMisses = Math.round(requests * (1 - hitRatio));
    const miss = Math.floor((index + 1) * targetMisses / requests) > Math.floor(index * targetMisses / requests);
    if (miss) return withNonce(base, `${runId}:miss:${index}`);
  }
  return base;
}

async function post(body) {
  const response = await fetch(url, { method: "POST", headers: { "content-type": "application/json" }, body });
  await response.arrayBuffer();
  if (!response.ok) throw new Error(`HTTP ${response.status}`);
  return response;
}

async function pool(items, worker) {
  let next = 0;
  await Promise.all(Array.from({ length: Math.min(concurrency, items.length) }, async () => {
    while (true) {
      const index = next++;
      if (index >= items.length) return;
      await worker(items[index], index);
    }
  }));
}

function percentile(values, quantile) {
  if (!values.length) return null;
  const ordered = [...values].sort((a, b) => a - b);
  return ordered[Math.max(0, Math.min(ordered.length - 1, Math.ceil(ordered.length * quantile) - 1))];
}

const versionUrl = new URL("/version", url).toString();
const serviceVersion = await (await fetch(versionUrl)).json();
if (scenario === "warm" || scenario === "mixed") await pool(payloads, post);
const scheduledStart = performance.now() + 250;
let inFlight = 0;
let next = 0;
await new Promise((resolve) => {
  const launch = () => {
    const now = performance.now();
    while (next < requests && inFlight < concurrency && scheduledStart + next * 1000 / rate <= now) {
      const index = next++;
      const scheduled = scheduledStart + index * 1000 / rate;
      inFlight++;
      post(bodyFor(index)).then((response) => {
        const done = performance.now();
        latencies.push(done - scheduled);
        const cache = response.headers.get("x-ecsae-cache") ?? "absent";
        cacheHeaders[cache] = (cacheHeaders[cache] ?? 0) + 1;
        const server = response.headers.get("x-ecsae-server-ms");
        if (server !== null) serverLatencies.push(Number(server));
      }).catch((error) => errors.push(String(error))).finally(() => {
        inFlight--;
        if (next === requests && inFlight === 0) resolve();
        else setImmediate(launch);
      });
    }
    if (next < requests) setTimeout(launch, Math.max(0, scheduledStart + next * 1000 / rate - performance.now()));
    else if (inFlight === 0) resolve();
  };
  launch();
});
const durationSeconds = (performance.now() - scheduledStart) / 1000;
const clientP95 = percentile(latencies, 0.95);
const serverP95 = percentile(serverLatencies, 0.95);
const report = {
  schema_version: "3",
  generated_at: new Date().toISOString(),
  generator: `Node ${process.version} built-in fetch; raw JSONL hit bodies`,
  model: "open-loop fixed-rate; latency = completion - scheduled_send",
  run_id: runId, scenario, expected_hit_fraction: scenario === "cold" ? 0 : scenario === "warm" ? 1 : hitRatio,
  payload_source: fs.realpathSync(payloadPath), distinct_base_payloads: payloads.length,
  payload_sha256: payloadSha256, service_version: serviceVersion,
  generator_and_server_colocated: true, server_workers: Number(argv["server-workers"] ?? 1),
  url, requested: requests, completed: latencies.length, errors: errors.length, error_examples: errors.slice(0, 10),
  target_rate_per_second: rate, achieved_rate_per_second: latencies.length / Math.max(durationSeconds, 1e-9),
  duration_seconds: durationSeconds, connections: concurrency, cache_headers: cacheHeaders,
  latency_ms: { min: Math.min(...latencies), mean: latencies.reduce((a, b) => a + b, 0) / latencies.length,
    p50: percentile(latencies, .5), p95: clientP95, p99: percentile(latencies, .99), max: Math.max(...latencies) },
  server_handler_latency_ms: { observations: serverLatencies.length,
    mean: serverLatencies.reduce((a, b) => a + b, 0) / serverLatencies.length,
    p50: percentile(serverLatencies, .5), p95: serverP95, p99: percentile(serverLatencies, .99), max: Math.max(...serverLatencies) },
  client_to_server_p95_ratio: clientP95 && serverP95 ? clientP95 / serverP95 : null,
  environment: { platform: `${os.platform()} ${os.release()}`, cpu_count: os.cpus().length, node: process.version },
};
fs.mkdirSync(path.dirname(output), { recursive: true });
fs.writeFileSync(output, JSON.stringify(report, null, 2) + "\n");
console.log(JSON.stringify({ cache: cacheHeaders, client_p95_ms: clientP95, server_p95_ms: serverP95,
  achieved_rps: report.achieved_rate_per_second, errors: errors.length }));
process.exitCode = errors.length ? 1 : 0;

File: bench/loadtest.py
"""Diverse-key open-loop HTTP benchmark with cold, mixed, and warm scenarios."""

from __future__ import annotations

import argparse
import asyncio
import json
import math
import platform
import statistics
import uuid
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

import httpx


def percentile(values: list[float], pct: float) -> float:
    ordered = sorted(values)
    index = max(0, min(len(ordered) - 1, math.ceil(pct * len(ordered)) - 1))
    return ordered[index]


def load_payloads(path: str | Path) -> list[dict[str, Any]]:
    source = Path(path)
    if source.suffix.lower() == ".jsonl":
        payloads = [json.loads(line) for line in source.read_text(encoding="utf-8").splitlines() if line]
    else:
        raw = json.loads(source.read_text(encoding="utf-8"))
        payloads = raw if isinstance(raw, list) else [raw]
    if not payloads:
        raise ValueError("at least one benchmark payload is required")
    return payloads


def with_nonce(payload: dict[str, Any], nonce: str) -> dict[str, Any]:
    return {
        **payload,
        "metadata": {**payload.get("metadata", {}), "benchmark_nonce": nonce},
    }


async def run(args: argparse.Namespace) -> dict[str, Any]:
    payloads = load_payloads(args.payloads)
    if not 0 <= args.hit_ratio <= 1:
        raise ValueError("hit_ratio must be between zero and one")
    run_id = args.run_id or str(uuid.uuid4())
    timeout = httpx.Timeout(args.timeout)
    limits = httpx.Limits(
        max_connections=args.connections,
        max_keepalive_connections=args.connections,
    )
    latencies: list[float] = []
    server_latencies: list[float] = []
    errors: list[str] = []
    cache_headers: dict[str, int] = {}
    semaphore = asyncio.Semaphore(args.connections)

    def payload_for(index: int) -> dict[str, Any]:
        base = payloads[index % len(payloads)]
        if args.scenario == "cold":
            return with_nonce(base, f"{run_id}:cold:{index}")
        if args.scenario == "mixed":
            # Bresenham-style spacing gives the requested aggregate fraction
            # for any run length without clustering all misses at the end.
            target_misses = round(args.requests * (1.0 - args.hit_ratio))
            is_miss = ((index + 1) * target_misses // args.requests) > (
                index * target_misses // args.requests
            )
            if is_miss:
                return with_nonce(base, f"{run_id}:miss:{index}")
        return base

    async with httpx.AsyncClient(timeout=timeout, limits=limits) as client:
        async def post(payload: dict[str, Any]) -> httpx.Response:
            async with semaphore:
                response = await client.post(args.url, json=payload)
                response.raise_for_status()
                return response

        prewarmed = 0
        if args.scenario in {"warm", "mixed"}:
            responses = await asyncio.gather(*(post(payload) for payload in payloads))
            prewarmed = len(responses)
        for index in range(args.warmup):
            await post(with_nonce(payloads[index % len(payloads)], f"{run_id}:warmup:{index}"))

        loop = asyncio.get_running_loop()
        start = loop.time() + 0.25

        async def one(index: int) -> None:
            scheduled = start + index / args.rate
            while (remaining := scheduled - loop.time()) > 0:
                await asyncio.sleep(remaining)
            try:
                response = await post(payload_for(index))
                done = loop.time()
                cache = response.headers.get("x-ecsae-cache", "absent")
                cache_headers[cache] = cache_headers.get(cache, 0) + 1
                latencies.append((done - scheduled) * 1000)
                server_value = response.headers.get("x-ecsae-server-ms")
                if server_value is not None:
                    server_latencies.append(float(server_value))
            except Exception as exc:
                errors.append(f"{type(exc).__name__}: {exc}")

        await asyncio.gather(*(one(index) for index in range(args.requests)))
        duration = loop.time() - start

    expected_hit_fraction = {
        "cold": 0.0,
        "mixed": args.hit_ratio,
        "warm": 1.0,
    }[args.scenario]
    return {
        "schema_version": "2",
        "generated_at": datetime.now(timezone.utc).isoformat(),
        "run_id": run_id,
        "model": "open-loop fixed-rate; latency = completion - scheduled_send",
        "scenario": args.scenario,
        "expected_hit_fraction": expected_hit_fraction,
        "payload_source": str(Path(args.payloads).resolve()),
        "distinct_base_payloads": len(payloads),
        "prewarmed_payloads": prewarmed,
        "url": args.url,
        "requested": args.requests,
        "completed": len(latencies),
        "errors": len(errors),
        "error_examples": errors[:10],
        "target_rate_per_second": args.rate,
        "achieved_rate_per_second": len(latencies) / max(duration, 1e-9),
        "duration_seconds": duration,
        "connections": args.connections,
        "excluded_warmup_requests": args.warmup,
        "cache_headers": cache_headers,
        "latency_ms": {
            "min": min(latencies) if latencies else None,
            "mean": statistics.fmean(latencies) if latencies else None,
            "p50": percentile(latencies, 0.50) if latencies else None,
            "p95": percentile(latencies, 0.95) if latencies else None,
            "p99": percentile(latencies, 0.99) if latencies else None,
            "max": max(latencies) if latencies else None,
        },
        "server_handler_latency_ms": {
            "observations": len(server_latencies),
            "mean": statistics.fmean(server_latencies) if server_latencies else None,
            "p50": percentile(server_latencies, 0.50) if server_latencies else None,
            "p95": percentile(server_latencies, 0.95) if server_latencies else None,
            "p99": percentile(server_latencies, 0.99) if server_latencies else None,
            "max": max(server_latencies) if server_latencies else None,
        },
        "client_to_server_p95_ratio": (
            percentile(latencies, 0.95) / percentile(server_latencies, 0.95)
            if latencies and server_latencies and percentile(server_latencies, 0.95) > 0
            else None
        ),
        "environment": {
            "python": platform.python_version(),
            "platform": platform.platform(),
        },
    }


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--url", default="http://127.0.0.1:8000/audit")
    parser.add_argument("--payloads", default="data/corpus/ctg-randomized-2000.jsonl")
    parser.add_argument("--scenario", choices=("cold", "mixed", "warm"), required=True)
    parser.add_argument("--hit-ratio", type=float, default=0.9)
    parser.add_argument("--requests", type=int, default=20_000)
    parser.add_argument("--rate", type=float, default=100)
    parser.add_argument("--connections", type=int, default=128)
    parser.add_argument("--warmup", type=int, default=100)
    parser.add_argument("--timeout", type=float, default=30)
    parser.add_argument("--run-id")
    parser.add_argument("--output", required=True)
    args = parser.parse_args()
    report = asyncio.run(run(args))
    target = Path(args.output)
    target.parent.mkdir(parents=True, exist_ok=True)
    target.write_text(
        json.dumps(report, indent=2, sort_keys=True) + "\n",
        encoding="utf-8",
        newline="\n",
    )
    print(json.dumps({"cache": report["cache_headers"], **report["latency_ms"]}, sort_keys=True))
    return 1 if report["errors"] else 0


if __name__ == "__main__":
    raise SystemExit(main())
